Added lifecycle hooks for enhanced workflow customization

mik047 opened the pull request:
Introduced the **lifecycle hooks** feature to the project, providing enhanced options for customizing and automating specific stages in the workflow. The primary focus of this feature is to allow developers to implement pre-defined hooks at critical stages of the application's lifecycle, facilitating greater flexibility and control.

#### **Key Changes:**
- Added a new `LifecycleHooks` interface to define hooks for key lifecycle events.
- Added hooks for `example validate` command

joelrosario committed:
Added initial version of lifecycle hooks

joelrosario committed:
Refactored the lifecycle hooks

joelrosario committed:
Refactored the lifecycle hook into a SAM interface

joelrosario committed:
Added a lifecycle hook

joelrosario committed:
Moved the lifecycle hook

joelrosario committed:
Updated hook implementation of AfterLoadingStaticExamplesHooks so that users of the hook don't need to know the underlying mechanism of how the hook is stored. Also now multiple users may invoke the hook.

mik047 committed:
bumped up version number as the previous one is taken by main

mik047 committed:
Bump the version from 2.7.8 to 2.7.15

mik047 committed:
Refactor imports and add lifecycle hook for examples in case spec file is provided

mik047 committed:
Add lifecycle hook for processing examples after validation for any duplication

mik047 committed:
Removing life cycle hook for stub command

mik047 committed:
Updating version from 2.7.15 to 2.7.8 based on latest released in main.

mik047 committed:
Remove unnecessary blank lines in ExamplesCommand.kt

joelrosario commented on core/src/main/kotlin/io/specmatic/core/lifecycle/LifecycleHooks.kt:
```diff
@@ -0,0 +1,5 @@
+package io.specmatic.core.lifecycle
+
+object LifecycleHooks {
+    var afterLoadingStaticExamples: AfterLoadingStaticExamplesHooks = AfterLoadingStaticExamplesHooks()
```
This should be a `val`, instead of `var`.

mik047 replied:
Done!

mik047 committed:
Added tests to ensure life cycle hook is getting called under various combination of `examples validate` command

mik047 committed:
Downgrade version from 2.7.8 to 2.7.6 as it was originally

mik047 committed:
Update version to 2.7.9

mik047 committed:
Merge branch 'main' into lifecycle_hooks

mik047 committed:
Fix order of implicit examples and external stubs in ExamplesCommandTest

mik047 committed:
Merge remote-tracking branch 'origin/lifecycle_hooks' into lifecycle_hooks

mik047 committed:
Remove unnecessary debug print statements from ExamplesCommandTest

mik047 committed:
Refactored example handling by consolidating file retrieval and lifecycle hook invocation

Also, to ensure the examples being validated and the one considered for duplication is the same list

mik047 committed:
Refactor ExamplesCommand to use immutable properties for directory options

mik047 committed:
Refactor import statements in API and ExamplesCommand to original state as those classes are no more required effectively for overall change

mik047 committed:
Change afterLoadingStaticExamples to use immutable properties

mik047 committed:
Change afterLoadingStaticExamples to use immutable property

joelrosario committed:
Minor test refactor
